Metagen update

leonidr-da opened the pull request.

leonidr-da committed:
Update README

leonidr-da committed:
XSD union's can have optionalAttributes

leonidr-da committed:
Add hie.yaml for Haskell Vscode integration

leonidr-da committed:
Remove unused import to quiet warning.

leonidr-da committed:
Add support for two more native Xml types.

leonidr-da committed:
Add extra Daml keyword to avoid

lucianojoublanc-da commented on tool-metagen/src/DA/Daml/GenDaml.hs:
```diff
@@ -32,12 +32,22 @@ ppImports = vcat . List.intersperse (text "") . map ppImport
+ppDecls useGhcPrimitives = vcat . List.intersperse (text "") . map (ppDecl useGhcPrimitives)
 
-ppDecl :: Show a => Decl a -> Doc
+derivingDecl :: Bool -> String -> Doc
```
I'm not sure what this is all about. Can you explain?

Regardless, I would clean this up a bit. For clarity I would remove this function, and inline it where it's used (line 63); it's confusing for instance that in line 50 you also have a `deriving (Eq, Ord Show)`. Also you'll be able to get rid of all those `_useGhcPrimitives` names at the signature level, which are unused in all but one declaration.

But I suspect I'm not understanding the purpose of this so the above may be invalid.

leonidr-da replied:
I am passing a flag down through these pretty print instructions to control whether we add a `deriving (Eq, Ord, Show)` line to our data types or to use custom instances based off of GHC primitives.

Agreed, the function makes more sense closer to the usage.

lucianojoublanc-da replied:
> to use custom instances based off of GHC primitives.

That's the bit I don't understand. Why do you need to do that?

leonidr-da replied:
To avoid recursion limits. For another project we were not able to upload the Xml contents because `Show` would break a (gRPC or Java, i think) recursion limit.

lucianojoublanc-da commented on tool-metagen/src/DA/XML/Schema/Convert.hs:
```diff
@@ -664,6 +664,7 @@ mkFieldType QName{..} =
         "any"                | isXsd -> Prim PrimText
 
         "decimal"            | isXsd -> Prim PrimDecimal
+        "double"             | isXsd -> Prim PrimDecimal
```
Going from floating point to fixed-point is probably a bad idea ... Are there any checks to ensure the conversion is sane?

leonidr-da replied:
The checks would occur at ingestion time (or better yet via a specific library designed for it such as `lib-integration-java`), but I think that for now (given competing priorities) storing it as a Decimal is the right approach. If we're being pedantic, and XML Decimal is not a Daml Decimal.

lucianojoublanc-da commented on tool-metagen/tx-target/.daml/package-database/1.13/settings:
```diff
@@ -0,0 +1 @@
+[("target arch","ArchUnknown"),("target os","OSUnknown"),("target word size","8"),("Unregisterised","YES"),("target has GNU nonexec stack","YES"),("target has .ident directive","YES"),("target has subsections via symbols","YES"),("cross compiling","NO")]
```
Did you deliberately check-in these files?

leonidr-da replied:
No, this was an error.

leonidr-da replied:
Apologies, I'll rebase the PR.

lucianojoublanc-da reviewed (commented):
I'm not certain whether we should be checking in the `tx-target` directory. It seems to be a package management concern, so typically I don't think this is is something you'd check-in.

I don't have enough Haskell knowledge to know whether this is considered good/bad practice, though. Can you get another opinion?

leonidr-da committed:
Add a flag to use GHC primitives for common typeclasses

there are Xml structures that result in too deeply nested and cause recursion failures with the default deriving functions.

leonidr-da commented:
@lucianojoublanc-da Removed errant files and moved function to local scope.

leonidr-da committed:
Sort in alphabetical order (within destination) conversion map.

leonidr-da committed:
Add conversions for unsigned byte and int.

leonidr-da commented:
@lucianojoublanc-da @georg-da Could we please move this along?

lucianojoublanc-da reviewed (approved):
Sorry I was on leave last week with no mobile access.

From my end this is ok now. the only thing I would consider is whether we need that flag for 'gch primitive' typeclass derivation. i.e. could we just replace all typeclass derivation to use this method, so as to reduce the complexity of having that extra flag? Or are there really two separate cases here - I just don't know enough about how this works to be able to comment intelligently.

leonidr-da commented:
Thanks for merging.

> could we just replace all typeclass derivation to use this method, so as to reduce the complexity of having that extra flag? 
I had the same thought. I think this is the most pragmatic approach to leave us options going forward, but agree that it is not ideal. I think that neither of our teams has the bandwidth at the moment to full evaluate the nuances and use cases.
